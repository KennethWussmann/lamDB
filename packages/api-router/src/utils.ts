import { isExecutableDefinitionNode, OperationDefinitionNode, OperationTypeNode, parse } from 'graphql';
import { Request, Response, sha1Hash } from '@lamdb/core';
import { Request as ExpressRequest, Response as ExpressResponse } from 'express';

export const graphQlErrorResponse = (message: string): Response => ({
  status: 400,
  body: JSON.stringify({
    data: null,
    errors: [
      {
        message,
      },
    ],
  }),
  headers: {
    'content-type': 'application/json',
  },
});

export const fromExpressRequest = (expressRequest: ExpressRequest): Request => ({
  method: expressRequest.method,
  body: expressRequest.body,
  headers: Object.fromEntries(
    Object.entries(expressRequest.headers).map(([key, value]) => [key, Array.isArray(value) ? value[0] : value ?? '']),
  ),
  path: expressRequest.path,
});

export const applyToExpressResponse = (response: Response, expressResponse: ExpressResponse) => {
  Object.entries(response.headers).forEach(([key, value]) => expressResponse.setHeader(key, value));
  expressResponse.status(response.status);
  expressResponse.send(response.body);
};

export const getOperationInfo = (
  request: Request,
): { name: string; type: OperationTypeNode; hash: string | undefined } | undefined => {
  if (!request.body) {
    return undefined;
  }

  let parsedDocument;
  try {
    parsedDocument = JSON.parse(request.body);
  } catch {
    return undefined;
  }

  const document = parse(parsedDocument.query, {
    noLocation: true,
  });

  const executableNodes = document.definitions.filter(isExecutableDefinitionNode);

  if (executableNodes.length === 0) {
    throw new Error('Document does not include executable nodes');
  }

  const operationNodes: OperationDefinitionNode[] = executableNodes
    .filter((node: any) => !!node?.operation)
    .map((node) => node as OperationDefinitionNode);

  if (operationNodes.length === 0) {
    throw new Error('Document does not include operation nodes');
  }
  if (operationNodes.length > 1) {
    throw new Error('Document includes multiple operation nodes: Only one is supported');
  }
  return {
    name: parsedDocument.operationName ?? operationNodes[0].name ?? '',
    type: operationNodes[0].operation,
    hash: request.body ? sha1Hash(request.body) : undefined,
  };
};
