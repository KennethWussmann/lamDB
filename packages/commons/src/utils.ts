import { createHash } from 'crypto';
import { access } from 'fs/promises';
import { Request, Response } from './requestResponse';
import { isExecutableDefinitionNode, OperationDefinitionNode, OperationTypeNode, parse } from 'graphql';

export const sha1Hash = (operation: string): string => {
  const hashsum = createHash('sha1');
  hashsum.update(operation);
  return hashsum.digest('hex').toString();
};

export const exists = async (file: string) => {
  try {
    await access(file);
    return true;
  } catch {
    return false;
  }
};

export const errorLog = (e: unknown): object => {
  const error = e as Error;
  return {
    error: {
      name: error.name,
      message: error.message,
      cause: error.cause,
      stack: error.stack,
    },
  };
};

export const graphQlErrorResponse = (message: string): Response => ({
  status: 400,
  body: JSON.stringify({
    data: null,
    errors: [
      {
        message,
      },
    ],
  }),
  headers: {
    'content-type': 'application/json',
  },
});

export const getOperationInfo = (request: Request): { name: string; type: OperationTypeNode; hash: string | undefined } | undefined => {
  if (!request.body) {
    return undefined;
  }

  let parsedDocument;
  try {
    parsedDocument = JSON.parse(request.body);
  } catch {
    return undefined;
  }

  const document = parse(parsedDocument.query, {
    noLocation: true,
  });

  const executableNodes = document.definitions.filter(isExecutableDefinitionNode);

  if (executableNodes.length === 0) {
    throw new Error('Document does not include executable nodes');
  }

  const operationNodes: OperationDefinitionNode[] = executableNodes.filter((node: any) => !!node?.operation).map((node) => node as OperationDefinitionNode);

  if (operationNodes.length === 0) {
    throw new Error('Document does not include operation nodes');
  }
  if (operationNodes.length > 1) {
    throw new Error('Document includes multiple operation nodes: Only one is supported');
  }
  return {
    name: parsedDocument.operationName ?? operationNodes[0].name ?? '',
    type: operationNodes[0].operation,
    hash: request.body ? sha1Hash(request.body) : undefined,
  };
};
