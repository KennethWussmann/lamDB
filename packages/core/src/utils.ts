import { createHash } from 'crypto';
import { access } from 'fs/promises';

export const sha1Hash = (operation: string): string => {
  const hashsum = createHash('sha1');
  hashsum.update(operation);
  return hashsum.digest('hex').toString();
};

export const exists = async (file: string) => {
  try {
    await access(file);
    return true;
  } catch {
    return false;
  }
};

export const errorLog = (e: unknown): object => {
  const error = e as Error;
  return {
    error: {
      name: error.name,
      message: error.message,
      cause: error.cause,
      stack: error.stack,
    },
  };
};

export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=5`;
