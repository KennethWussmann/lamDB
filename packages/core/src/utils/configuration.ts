// istanbul ignore file
import { config } from 'dotenv';

config();
export class LamDBConfiguration {
  get migrationEngineBinaryPath() {
    return assertEnv('MIGRATION_ENGINE_BINARY_PATH');
  }
  get migrationEngineForceMigration() {
    return assertEnv('MIGRATION_ENGINE_FORCE_MIGRATION', 'false').toLowerCase() === 'true';
  }
  get queryEngineLibraryPath() {
    return assertEnv('QUERY_ENGINE_LIBRARY_PATH');
  }
  get databasePath() {
    return assertEnv('DATABASE_PATH');
  }
  get prismaSchemaPath() {
    return assertEnv('PRISMA_SCHEMA_PATH');
  }
  get operationOptimization() {
    return assertEnv('OPERATION_OPTIMIZATION', 'false').toLowerCase() === 'true';
  }

  get server() {
    return new LamDBServerConfiguration();
  }
}

export class LamDBServerConfiguration {
  get allowCors() {
    return assertEnv('ALLOW_CORS', 'false').toLowerCase() === 'true';
  }
  get proxyPort() {
    return parseInt(assertEnv('PROXY_PORT', '4000'), 10);
  }
}

const assertEnv = (name: string, defaultValue: string | undefined = undefined): string => {
  const value = process.env[name] ?? defaultValue;
  if (!value) {
    throw new Error(`Environment variable ${name} not set, but required`);
  }
  return value;
};
