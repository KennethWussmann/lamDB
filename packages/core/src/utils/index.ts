export * from './utils';
export * from './configuration';
