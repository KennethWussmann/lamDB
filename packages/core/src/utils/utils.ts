export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=3&connection_limit=1`;
