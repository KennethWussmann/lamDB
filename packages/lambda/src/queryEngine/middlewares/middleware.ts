import { Logger } from 'winston';
import { Request, Response } from '../../requestResponse';
import { QueryEngine } from '../queryEngine';

export type MiddlewareContext = {
  request: Request;
  queryEngine: QueryEngine;
  logger: Logger;
  host: string;
  port: number;
};
export type MiddlewareNextFunction = (context: MiddlewareContext) => Promise<Response>;
export type QueryEngineProxyMiddleware = (
  context: MiddlewareContext,
  next: MiddlewareNextFunction,
) => Promise<Response> | Response;

export const executeMiddlewares = async (
  context: MiddlewareContext,
  middlewares: QueryEngineProxyMiddleware[],
  execute: (context: MiddlewareContext) => Promise<Response>,
): Promise<Response> => {
  const exec = async (context: MiddlewareContext, middlewares: QueryEngineProxyMiddleware[]): Promise<Response> => {
    if (!middlewares.length) {
      return await execute(context);
    }
    const middleware = middlewares[0];
    const response = await middleware(context, async (ctx) => await exec(ctx, middlewares.slice(1)));
    return response;
  };
  return await exec(context, middlewares);
};
