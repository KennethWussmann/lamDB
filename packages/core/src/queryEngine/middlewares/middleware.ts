import { Logger } from 'winston';
import { Request, Response } from '../../utils';
import { QueryEngine } from '../queryEngine';

export type MiddlewareContext = {
  request: Request;
  queryEngine: QueryEngine;
  logger: Logger;
};
export type MiddlewareNextFunction = (context: MiddlewareContext) => Promise<Response>;
export type QueryEngineProxyMiddleware = (
  context: MiddlewareContext,
  next: MiddlewareNextFunction,
) => Promise<Response> | Response;

export const executeMiddlewares = async (
  context: MiddlewareContext,
  middlewares: QueryEngineProxyMiddleware[],
  execute: (context: MiddlewareContext) => Promise<Response>,
): Promise<Response> => {
  const exec = async (context: MiddlewareContext, middlewares: QueryEngineProxyMiddleware[]): Promise<Response> => {
    if (!middlewares.length) {
      return await execute(context);
    }
    const middleware = middlewares[0];
    const response = await middleware(context, async (ctx) => await exec(ctx, middlewares.slice(1)));
    return response;
  };

  if (middlewares.length === 0) {
    return await execute(context);
  }

  return await exec(context, middlewares);
};

export interface Middleware {
  handle: QueryEngineProxyMiddleware;
}
