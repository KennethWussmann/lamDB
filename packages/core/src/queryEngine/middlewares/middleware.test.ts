import { testQuery } from '../../../test/testOperation';
import { getTestQueryEngine } from '../../../test/testQueryEngine';
import { operationToRequest } from '../../../test/utils';
import { Response, createLogger } from '@lamdb/commons';
import { executeMiddlewares, MiddlewareContext, QueryEngineProxyMiddleware } from './middleware';

let context: MiddlewareContext;

describe('middleware', () => {
  beforeEach(() => {
    context = {
      queryEngine: getTestQueryEngine(),
      logger: createLogger({ name: 'test' }),
      request: operationToRequest(testQuery),
    };
  });

  it('executes middlewares in order', async () => {
    const end = (context: MiddlewareContext): Promise<Response> =>
      Promise.resolve({
        status: 201,
        body: `Test Body for request ${context.request.method} ${context.request.path}`,
        headers: {
          'content-type': 'application/json',
        },
      });
    const middlewareOne: QueryEngineProxyMiddleware = (context, next) =>
      next({
        ...context,
        request: {
          ...context.request,
          path: '/test',
        },
      });
    const middlewareTwo: QueryEngineProxyMiddleware = (context, next) =>
      next({
        ...context,
        request: {
          ...context.request,
          path: `${context.request.path}/two`,
        },
      });

    const response = await executeMiddlewares(context, [middlewareOne, middlewareTwo], end);

    expect(response).toMatchSnapshot();
  });
});
