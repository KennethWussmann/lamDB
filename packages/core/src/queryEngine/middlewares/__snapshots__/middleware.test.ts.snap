// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware executes middlewares in order 1`] = `
{
  "body": "Test Body for request POST /test/two",
  "headers": {
    "content-type": "application/json",
  },
  "status": 201,
}
`;
